internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+	// Start time first.
+	notBeforeSort := t.Certificate.NotBefore.Compare(other.Certificate.NotBefore)
+	if notBeforeSort != 0 {
+		return notBeforeSort
+	}
+	// End time next.
+	notAfterSort := t.Certificate.NotAfter.Compare(other.Certificate.NotAfter)
+	if notAfterSort != 0 {
+		return notAfterSort
+	}
+	// Fingerprint last.
+	return strings.Compare(t.Fingerprint(), other.Fingerprint())
+}
+
+// Fingerprint returns an unpadded base32 encoded SHA512/384 hash of the raw
+// x509 certificate.
+func (t TLSKeyPair) Fingerprint() string {
+	return t.fingerprint
+}
+
+// TLSCertificate returns a TLS certificate suitable for configuring a
+// TLS based listener.
+func (t TLSKeyPair) TLSCertificate() (*tls.Certificate, error) {
+	pemCert, err := t.PEMCertificate()
+	if err != nil {
+		return nil, err
+	}
+	pemKey, err := t.pemKey()
+	if err != nil {
+		return nil, err
+	}
+	tlsCert, err := tls.X509KeyPair(pemCert, pemKey)
+	if err != nil {
+		return nil, err
+	}
+	return &tlsCert, err
+}
+
+// PEMCertificate converts the X509 certificate to a PEM encoded byte slice,
+// useful for sending to daemon API clients.
+func (t TLSKeyPair) PEMCertificate() ([]byte, error) {
+	pemBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: t.Certificate.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert certificate to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// pemKey returns a PKCS8 encoded private key in PEM format.
+func (t TLSKeyPair) pemKey() ([]byte, error) {
+	privateBytes, err := x509.MarshalPKCS8PrivateKey(t.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert key to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// toPEMBytes converts the private key and x509 certificate into a byte slice
+// that contains two PEM blocks, ready for secure storage on disk.
+func (t TLSKeyPair) toPEMBytes() ([]byte, error) {
+	privateBytes, err := x509.MarshalPKCS8PrivateKey(t.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateBytes,
+	}
+	pemCertificateBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: t.Certificate.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert key to PEM: %w", err)
+	}
+	if err := pem.Encode(&pemBuffer, pemCertificateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert x509 certificate to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// fromPEMFile loads a PKCS8 private key and x509 certificate
+// from a PEM formatted file.
+func (t *TLSKeyPair) fromPEMFile(path string) error {
```

Comment:
```suggestion
func loadKeyPair(path string) (*TLSKeyPair, error) {
```

Reply:
PR reworked. N/A for current PR. Resolving.